python/rainbow/simulators/inverse_kinematics/api.py

```diff
@@ -0,0 +1,555 @@
+    end-effector positions in world-space given the cufrrent joint-angle values.
+
+    :param chains:    The IK chains defining the end-effectors.
+    :param skeleton:  The IK skeleton.
+    :return:          A numpy matrix with the value of the Jacobian.
+    """
+    num_angles = len(skeleton.bones) * 3
+    num_coords = len(chains) * 3
+    J = np.zeros((num_coords, num_angles), dtype=np.float64)
+    row_offset = 0
+    for chain in chains:
+        e = get_end_effector(chain, skeleton)
+        for idx in chain.bones:
+            bone = skeleton.bones[idx]
+            q_parent = Q.identity()
+            if bone.has_parent():
+                q_parent = skeleton.bones[bone.parent].q_wcs
+
+            # In the 5 lines of code below the ZYZ Euler angles are hardwired into the code.
+            q_alpha = bone.get_rotation_alpha()
+            q_alpha_beta = Q.prod(q_alpha, bone.get_rotation_beta())
+
+            u = Q.rotate(q_parent, bone.get_axis_alpha())
+            v = Q.rotate(q_parent, Q.rotate(q_alpha, bone.get_axis_beta()))
+            w = Q.rotate(q_parent, Q.rotate(q_alpha_beta, bone.get_axis_gamma()))
+
+            delta_p = e - bone.t_wcs
+
+            J_alpha = V3.cross(u, delta_p)
+            J_beta = V3.cross(v, delta_p)
+            J_gamma = V3.cross(w, delta_p)
+
+            J[row_offset: row_offset + 3, idx * 3 + 0] = J_alpha
+            J[row_offset: row_offset + 3, idx * 3 + 1] = J_beta
+            J[row_offset: row_offset + 3, idx * 3 + 2] = J_gamma
+        row_offset += 3
+    return J
+
+
+def compute_hessian(chains, skeleton, J):
+    """
+    This function compute the Hessian of the IK objective function wrt all the joint angles.
+
+    This implementation is based on an analytical solution for computing the gradient and is hence much faster than
+    a numerical approximation method.
+
+    :param chains:       The IK chains of the IK problem.
+    :param skeleton:     The IK skeleton (basically storing the joint angle values)
+    :param J:            The Jacobian matrix of the forward kinematics functions. Should be computed
+                         using the function compute_jacobian.
+    :return:             A numpy matrix holding the Hessian value.
+    """
+    H = np.dot(np.transpose(J), J)
+    row_offset = 0
+    for chain in chains:
+        e = get_end_effector(chain, skeleton)
+        r = chain.goal - e
+        for k in chain.bones:
+            for h in chain.bones:
+                if h > k:
+                    continue
+
+                Bh = skeleton.bones[h]
+
+                q_parent = Q.identity()
+                if Bh.has_parent():
+                    q_parent = skeleton.bones[Bh.parent].q_wcs
+
+                # In the 7 lines of code below the ZYZ Euler angles are hardwired into the code.
+                q_alpha = Bh.get_rotation_alpha()
+                q_beta = Bh.get_rotation_beta()
+                q_gamma = Bh.get_rotation_gamma()
```

Comment:
Is this unused?

Reply:
q_gamma = Bh.get_rotation_gamma()

Reply:
Yes, but it is left in the code for didactic purposes. An optimized implementation would likely out comment this line.